proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1160 @@
+                        "fieldNumber": 1
+                    },
+                    "amount": {
+                        "dataType": "sint64",
+                        "fieldNumber": 2
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+The verification and execution of this transaction is specified in [LIP 0023][LIP-0023-voteTrs]. 
+This specification is followed to implement the vote command with the following additional point:
+
+* when executing a self-vote (a vote with the `delegateAddress` equal to the address derived from the transaction public key), 
+  modify the `delegateStore(delegateAddress).selfVotes` property and the `voterStore(delegateAddress).sentVotes` entry corresponding to `delegateAddress` identically.
+
+
+#### Unlock
+
+Transactions executing this command have:
+
+* `moduleID = MODULE_ID_DPOS`
+* `commandID = COMMAND_ID_UNLOCK`
+
+
+##### Parameters
+
+The `params` property of unlock transactions is empty.
+
+
+#### Verification
+
+Unlock transactions do not trigger specific verifications.
+
+
+#### Execution
+
+When executing an unlock transaction `trs`, the following is done:
+
+```python
+senderAddress = address corresponding to trs.senderPublicKey
+height = height of the block including trs
+
+for each unlockObject in voterStore(senderAddress).pendingUnlocks:
+    if (hasWaited(unlockObject, senderAddress, height) 
+        and not isPunished(unlockObject, senderAddress, height)
+        and isCertificateGenerated(unlockObject)):
+        delete unlockObject from voterStore(senderAddress).pendingUnlocks
+        token.unlock(senderAddress, MODULE_ID_DPOS, TOKEN_ID_DPOS, unlockObject.amount)
+```
+
+The definition and rationale for the `isCertificateGenerated` function is part of [LIP "Introduce unlocking condition for incentivizing certificate generation"][LIP-incentivizeCertificateGeneration]. 
+The `hasWaited` and `isPunished` functions are defined below and are rationalized in [LIP 0023][LIP-0023-unlockRationale] and [LIP 0024][LIP-0024-rationale] respectively.
+```python
+hasWaited(unlockObject, senderAddress, height):
+    if unlockObject.delegateAddress == senderAddress:
+        # this is a self-unvote
+        delayedAvailability = 260,000
+    else:
+        delayedAvailability = 2000
+
+    if height - unlockObject.unvoteHeight < delayedAvailability:
+        return False
+    else:
+        return True
+```
+
+```python
+isPunished(unlockObject, senderAddress, height):
+    if delegateStore(unlockObject.delegateAddress).pomHeights is empty:
+        return false
+    else:
+        let lastPomHeight be the last element of delegateStore(unlockObject.delegateAddress).pomHeights 
+        # lastPomHeight is also its largest element of the pomHeights array
+        
+        if unlockObject.address == senderAddress:
+            # this is a self-unvote
+            if height – lastPomHeight < 780,000 and lastPomHeight < unlockObject.unvoteHeight + 260,000:
```

Comment:
couldnt this return true for `height < lastPomHeight`? I understand the normal use would be to set height to the block height at which the unlocking transaction is included, but there is no check that, say `height > unlockObject.unvoteHeight`

Reply:
I'm not sure that I see your concern. There is indeed not check that `height > unlockObject.unvoteHeight`, but the `unvoteHeight` is set when you unvote, which must happen before you unlock. Otherwise, the unlockObject does not exist.
Maybe the implementation can add checks for those non-existent cases, but I don't think they need to be in the LIP.

Reply:
ah sorry, I somehow missed this is only an internal function at this point